core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4

```diff
@@ -119,35 +119,44 @@ statement
         (WITH ADMIN grantor)?
         (IN catalog=identifier)?                                       #createRole
     | DROP ROLE name=identifier (IN catalog=identifier)?               #dropRole
-    | GRANT
-        roles
-        TO principal (',' principal)*
-        (WITH ADMIN OPTION)?
-        (GRANTED BY grantor)?
-        (IN catalog=identifier)?                                       #grantRoles
-    | REVOKE
-        (ADMIN OPTION FOR)?
-        roles
-        FROM principal (',' principal)*
-        (GRANTED BY grantor)?
-        (IN catalog=identifier)?                                       #revokeRoles
-    | SET ROLE (ALL | NONE | role=identifier)
-        (IN catalog=identifier)?                                       #setRole
-    | GRANT
-        (privilege (',' privilege)* | ALL PRIVILEGES)
-        ON (SCHEMA | TABLE)? qualifiedName
-        TO grantee=principal
-        (WITH GRANT OPTION)?                                           #grant
+    | GRANT (
+        (
```

Comment:
Have you tried using two separate labels for each of the branches instead of bundling them under `#genericGrant` ?

Reply:
I did try.  They problem I saw an error in the IDE when was #grant inside the required outer parens.  AFAICT #tags are only allowed at the top level.

Reply:
Then, I would try the following:

```
| GRANT privilegeOrRole (',' privilegeOrRole)*
            TO principal (',' principal)*
            (WITH ADMIN OPTION)?
            (GRANTED BY grantor)?
            (IN catalog=identifier)?   #grantRole
| GRANT ((privilegeOrRole (',' privilegeOrRole)*) | ALL PRIVILEGES)
            ON grantObject
            TO principal
            (WITH GRANT OPTION)? #grantPrivilege
```

Reply:
Awesome, @martint, that worked and all parser tests pass!

I pushed the improved version.